test/lit/gc-read-write-effects.wast

```diff
@@ -0,0 +1,73 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py and should not be edited.
+
+;; Check that writing a struct field is not reordered with reading the same
+;; struct field.
+
+;; RUN: wasm-opt -all -O1 %s -S -o - | filecheck %s
+
+(module
+ (type $A (struct
+   (field $i (mut i32))
+   (field $b (mut (ref null $B)))
+ ))
+ (type $B (struct
+   (field $i (mut i32))
+   (field $c (mut (ref null $C)))
+   (field $b (mut (ref null $B)))
+ ))
+ (type $C (struct
+   (field (mut i32))
+   (field (mut i64))
+ ))
+ (export "f" (func $f))
+
+ ;; Check that this:
+ ;;
+ ;;   c = a.b.c
+ ;;   a.b = a.b.b
+ ;;   return c
+ ;;
+ ;; Is not turned into this:
+ ;;
+ ;;   a.b = a.b.b
+ ;;   return a.b.c
+ ;;
+ ;; CHECK:      (func $f (param $0 (ref null $A)) (result (ref null $C))
+ ;; CHECK-NEXT:  (local $1 (ref null $C))
+ ;; CHECK-NEXT:  (local.set $1
+ ;; CHECK-NEXT:   (struct.get $B $c
+ ;; CHECK-NEXT:    (struct.get $A $b
+ ;; CHECK-NEXT:     (local.get $0)
+ ;; CHECK-NEXT:    )
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (struct.set $A $b
+ ;; CHECK-NEXT:   (local.get $0)
+ ;; CHECK-NEXT:   (struct.get $B $b
+ ;; CHECK-NEXT:    (struct.get $A $b
+ ;; CHECK-NEXT:     (local.get $0)
+ ;; CHECK-NEXT:    )
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (local.get $1)
+ ;; CHECK-NEXT: )
+ (func $f (param $0 (ref null $A)) (result (ref null $C))
+  (local $1 (ref null $C))
+  (local.set $1
+   (struct.get $B 1
+    (struct.get $A 1
+     (local.get $0)
+    )
+   )
+  )
+  (struct.set $A 1
+   (local.get $0)
+   (struct.get $B 2
+    (struct.get $A 1
+     (local.get $0)
+    )
+   )
+  )
+  (local.get $1)
+ )
+)
```

Comment:
This testcase can be simplified to
```wat
(module
 (type $A (struct (field (mut i32))))
 (func "test" (param $x (ref null $A)) (result i32)
  (local $y i32)
  (local.set $y
   (struct.get $A 0
    (local.get $x)
   )
  )
  (struct.set $A 0
   (local.get $x)
   (i32.const 10)
  )
  (local.get $y)
 )
)
```
And you just need to run `--simplify-locals` and not all of `-O1`.